From 9e2d07f38d904b3334f2e48b7e3b00a88a4241d7 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@st.com>
Date: Mon, 29 Sep 2025 21:28:56 +0700
Subject: [PATCH 02/12] ta: pkcs11: be flexible on RSA private key optional
 attributes

Allow RSA private key to partially store the RSA private key optional
attribute without facing complains from the GP TEE API.

On one hand, in the PKCS#11 specification, RSA private key attributes
CKA_PRIME_1, CKA_PRIME_2, CKA_EXPONENT_1, CKA_EXPONENT_2 and
CKA_COEFFICIENT are optional and the spec does not add much constraint
on their presence.

In the other hand, the GP TEE Internal Core API request that these
5 optional attribute are all present or not present at all.

As a trade-off, allow PKCS#11 client to partially provide them but do
not
load them into the TEE object unless they are all present.

Fixes: 3dc4089 ("ta: pkcs11: correct RSA keys extended attributes
sanitation")
Fixes: OP-TEE#5418
Fixes: OP-TEE#7520
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>
Signed-off-by: Dien Pham <dien.pham.ry@renesas.com>
---
 ta/pkcs11/src/processing_rsa.c | 27 ++++++++++++++-------------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/ta/pkcs11/src/processing_rsa.c b/ta/pkcs11/src/processing_rsa.c
index 94eabbfa7..ced963a2d 100644
--- a/ta/pkcs11/src/processing_rsa.c
+++ b/ta/pkcs11/src/processing_rsa.c
@@ -511,7 +511,6 @@ enum pkcs11_rc load_tee_rsa_key_attrs(TEE_Attribute **tee_attrs,
 	TEE_Attribute *attrs = NULL;
 	size_t count = 0;
 	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
-	void *a_ptr = NULL;
 
 	assert(get_key_type(obj->attributes) == PKCS11_CKK_RSA);
 
@@ -559,16 +558,12 @@ enum pkcs11_rc load_tee_rsa_key_attrs(TEE_Attribute **tee_attrs,
 		if (count != 3)
 			break;
 
-		/* If pre-computed values are present load those */
-		rc = get_attribute_ptr(obj->attributes, PKCS11_CKA_PRIME_1,
-				       &a_ptr, NULL);
-		if (rc != PKCS11_CKR_OK && rc != PKCS11_RV_NOT_FOUND)
-			break;
-		if (rc == PKCS11_RV_NOT_FOUND || !a_ptr) {
-			rc = PKCS11_CKR_OK;
-			break;
-		}
-
+		/*
+		 * If pre-computed values are present load those
+		 * but only if they are all present since the GP TEE
+		 * specification expects either the 5 are present
+		 * or none is present.
+		 */
 		if (pkcs2tee_load_attr(&attrs[count], TEE_ATTR_RSA_PRIME1, obj,
 				       PKCS11_CKA_PRIME_1))
 			count++;
@@ -589,9 +584,15 @@ enum pkcs11_rc load_tee_rsa_key_attrs(TEE_Attribute **tee_attrs,
 				       obj, PKCS11_CKA_COEFFICIENT))
 			count++;
 
-		if (count == 8)
-			rc = PKCS11_CKR_OK;
+		/*
+		 * As mentionned above, if at least 1of these optional
+		 * attribute is missing, keep only the 3 first already
+		 * loaded above.
+		 */
+		if (count != 8)
+			count = 3;
 
+		rc = PKCS11_CKR_OK;
 		break;
 
 	default:
-- 
2.34.1

